src/persist/src/file.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! In-memory implementations for testing and benchmarking.
+
+use std::convert::TryInto;
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Seek, SeekFrom, Write as StdWrite};
+use std::sync::{Arc, Mutex};
+
+use crate::error::Error;
+use crate::mem::MemSnapshot;
+use crate::persister::{Meta, Write};
+
+/// A naive implementation of [Write] and [Meta] backed by a file.
+#[derive(Clone, Debug)]
+pub struct FileStream {
+    dataz: Arc<Mutex<File>>,
+}
+
+impl FileStream {
+    /// Create a new FileStream
+    pub fn new(name: String) -> Self {
+        // TODO: could not get truncate to work here.
+        let file = OpenOptions::new()
+            .write(true)
+            .append(true)
+            .create_new(true)
+            .read(true)
+            .open(name)
+            .expect("failed to create file");
+        FileStream {
+            dataz: Arc::new(Mutex::new(file)),
+        }
+    }
+}
+
+impl Write for FileStream {
+    fn write_sync(&mut self, updates: &[((String, String), u64, isize)]) -> Result<(), Error> {
+        let mut buf = Vec::new();
+
+        for ((key, val), ts, diff) in updates.iter() {
+            let key_bytes = key.as_bytes();
+            let key_bytes_len = key_bytes.len() as u64;
+            let val_bytes = val.as_bytes();
+            let val_bytes_len = val_bytes.len() as u64;
+
+            buf.extend(&key_bytes_len.to_le_bytes());
+            buf.extend(key_bytes);
+            buf.extend(&val_bytes_len.to_le_bytes());
+            buf.extend(val_bytes);
+            buf.extend(&ts.to_le_bytes());
+            buf.extend(&(*diff as i64).to_le_bytes());
+        }
+        let mut guard = self.dataz.lock()?;
+        guard.write_all(&buf).expect("failed to write file");
+        guard.sync_all().expect("failed to sync file");
+        Ok(())
+    }
+}
+
+impl Meta for FileStream {
+    type Snapshot = MemSnapshot;
+
+    fn snapshot(&self) -> Result<Self::Snapshot, Error> {
+        let mut bytes = Vec::new();
+        let mut dataz = Vec::new();
+
+        let mut guard = self.dataz.lock()?;
+        guard
+            .seek(SeekFrom::Start(0))
+            .expect("failed to seek to beginning");
+        guard.read_to_end(&mut bytes).expect("failed to read file");
+        // Unlock the mutex
+        drop(guard);
+        let mut bytes_remaining = bytes.as_slice();
+
+        while bytes_remaining.len() > 0 {
```

Comment:
`&[u8]` implements `std::io::Read`, so you can replace all the `split_at` and `remainder` stuff below with something like this:

```
let bytes = &bytes[..];
let mut len_buf = [0u8; 8];
while !bytes.is_empty() {
    bytes.read_exact(&mut len_buf[..])?;
    let key_len = u64::from_le_bytes(len_buf);
    ...
```

Reply:
tried to poke at this for a bit - I think you'll still need to use split_at for the actual variable length byte slices for the key and val (or something equivalent like index arithmetic into subslice selection?)

not sure if I'm missing something here.

Reply:
Is it not something like this? I didn't think about this closely, so it's easily possible that I'm the one missing something.

```
let key = Vec::new(key_len);
bytes.read_exact(&mut key[..])?;
```